posts/2018-04-13-rethinking-monaderror.md

```diff
@@ -0,0 +1,465 @@
+
+Instead, since I don't really like duplication and the fact that we're not going to deprecate `MonadError` overnight, I decided to extend `MonadBlunder` from `MonadError` for the `F` type, to get access to the `raiseError` function.
+If `raiseError` and `handleErrorWith` were instead separated into separate type classes (as is currently the case in the PureScript prelude), we could extend only the `raiseError` part.
+This also allows us to define laws that our counterparts of functions like `attempt` and `ensure` are consistent with the ones defined on `MonadError`.
+So the type signature now looks like this (expressed in Haskell, since it's easier on the eyes):
+```haskell
+class (MonadError f e, Monad g) => MonadBlunder f g e | f -> e, f -> g, g -> f where
+  ...
+```
+
+In Scala, we can't express this as nicely, so we're going to have to use something close to the `cats-mtl` encoding:
+
+```scala
+trait MonadBlunder[F[_], G[_], E] {
+  val monadErrorF: MonadError[F, E]
+  val monadG: Monad[G]
+  
+  ...
+}
+```
+
+Now since this means that any instance of `MonadBlunder` will also have an instance of `MonadError` on `F`, we might want to rename the functions we've got so far.
+Here's a complete definition of what we've come up with with `raiseError` removed and `handleErrorWith` renamed to `handleBlunderWith`:
+
+```scala
+trait MonadBlunder[F[_], G[_], E] {
+  val monadErrorF: MonadError[F, E]
+  val monadG: Monad[G]
+
+  def handleBlunderWith[A](fa: F[A])(f: E => G[A]): G[A]
+}
+```
+
+Now let us go back to defining more derived functions for `MonadBlunder`.
+The easiest probably being `handleError`, so let's see if we can come up with a good alternative:
+
+```scala
+trait MonadBlunder[F[_], G[_], E] {
+  ...
+
+  def handleBlunder[A](fa: F[A])(f: E => A): G[A] = 
+    handleBlunderWith(fa)(f andThen (_.pure[G]))
+}
+```
+
+This one is almost exactly like `handleBlunderWith`, but takes a function from `E` to `A` instead of to `G[A]`. We can easily reuse `handleBlunderWith` by using `pure` to go back to `E => G[A]`.
+
+Next another function that's really useful is `attempt`.
+Our alternative, let's call it `endeavor` for now, should return a value in `G` instead, which doesn't have a `MonadError` instance and therefore can not make any additional calls to `endeavor`:
+
+
+```scala
+trait MonadBlunder[F[_], G[_], E] {
+  ...
+
+  def endeavor[A](fa: F[A]): G[Either[E, A]] =
+    handleBlunder(fa.map(Right(_)))(Left(_))
+}
+```
+
+The implementation is fairly straightforward as well, we just handle all the errors my lifting them into the left side of an `Either` and map successful values to the right side of `Either`.
+
+Next, let's look at the dual to `attempt`, called `rethrow` in Cats. 
+For `MonadError` it turns an `F[Either[E, A]]` back into an `F`, but we're going to use our unexceptional type again:
+
+```scala
+trait MonadBlunder[F[_], G[_], E] {
+  ...
+
+  def absolve[A](fa: F[A]): G[Either[E, A]] = ???
```

Comment:
Wrong signature (the correct one is used below)

Reply:
Woops, thank you!